In [1]:
import pandas as pd

In [2]:
file = r"C:\Users\JAMIU\Desktop\Alt_school\Karatu Semester\Datasets\crime_data_chicago.csv"
crime_chic = pd.read_csv(file)

In [3]:
crime_chic.shape

(2278726, 23)

In [4]:
crime_chic.head(40)

Unnamed: 0        ID Case Number                    Date  \
0            0   6407111    HP485721  07/26/2008 02:30:00 PM   
1            1  11398199    JB372830  07/31/2018 10:57:00 AM   
2            2   5488785    HN308568  04/27/2007 10:30:00 AM   
3            3  11389116    JB361368  07/23/2018 08:55:00 AM   
4            4  12420431    JE297624  07/11/2021 06:40:00 AM   
5            5   1699235     G498287  08/21/2001 12:00:00 AM   
6            6   5061155    HM660983  10/14/2006 10:00:00 PM   
7            7   9876456    HX527438  12/02/2014 11:48:00 AM   
8            8   7582927    HS386492  06/30/2010 01:00:00 AM   
9            9  10566046    HZ313634  06/18/2016 11:15:00 PM   
10          10   9466303    HX118914  01/18/2014 09:00:00 AM   
11          11   6577203    HP649357  10/26/2008 05:00:00 PM   
12          12   8373423    HT607627  11/19/2011 09:00:00 AM   
13          13   3298634    HK328238  04/25/2004 04:00:00 PM   
14          14   8699049    HV375446  07/10/2012 07:59:00 AM   
15          15   9522880    HX178073  03/09/2014 12:30:00 PM   
16          16  10135550    HY318953  06/28/2015 03:50:00 AM   
17          17   3447250    HK514345  07/23/2004 12:00:00 AM   
18          18   2681993    HJ283703  04/06/2003 03:00:00 PM   
19          19   1932758    HH104177  01/02/2002 05:00:00 PM   
20          20   7315747    HS119810  01/14/2010 01:00:00 AM   
21          21   2409901    HH725517  10/19/2002 09:15:00 PM   
22          22   2265520    HH543376  07/29/2002 11:23:05 AM   
23          23   6593888    HP646969  10/23/2008 08:00:00 PM   
24          24  10473781    HZ213126  04/04/2016 02:34:00 PM   
25          25   3376825    HK428100  06/13/2004 06:00:00 AM   
26          26  11143513    JA503877  11/08/2017 01:10:00 PM   
27          27   3682751    HK783218  09/05/2004 06:00:00 PM   
28          28   9365306    HW508438  10/26/2013 01:45:00 PM   
29          29   7763508    HS571664  09/30/2010 09:00:00 PM   
30          30   1496484     G223033  03/20/2001 12:01:00 AM   
31          31   2718013    HJ347909  05/06/2003 10:30:00 PM   
32          32  12037807    JD217598  04/26/2020 01:55:00 PM   
33          33  12592835    JF113360  01/15/2022 03:59:00 AM   
34          34  12252122    JD467371  12/19/2020 12:00:00 AM   
35          35   7712225    HS518443  09/16/2010 12:00:00 PM   
36          36  10348890    HY539692  12/16/2015 05:00:00 PM   
37          37   7533272    HS335938  05/31/2010 05:00:00 PM   
38          38  11366391    JB331365  07/01/2018 11:55:00 PM   
39          39   3332372    HK373656  05/18/2004 04:30:00 PM   

                       Block  IUCR         Primary Type  \
0       085XX S MUSKEGON AVE  1320      CRIMINAL DAMAGE   
1          092XX S ELLIS AVE  143C    WEAPONS VIOLATION   
2          062XX N TRIPP AVE  0610             BURGLARY   
3         0000X N KEELER AVE  0560              ASSAULT   
4        016XX W HARRISON ST  051A              ASSAULT   
5              003XX W 28 PL  0810                THEFT   
6        006XX S CENTRAL AVE  0320              ROBBERY   
7        043XX W POTOMAC AVE  1811            NARCOTICS   
8         032XX W PIERCE AVE  0910  MOTOR VEHICLE THEFT   
9     021XX N CALIFORNIA AVE  1811            NARCOTICS   
10       031XX W WARREN BLVD  2014            NARCOTICS   
11        058XX S KEDZIE AVE  0486              BATTERY   
12     014XX N ELK GROVE AVE  0620             BURGLARY   
13        076XX S HALSTED ST  2830        OTHER OFFENSE   
14       042XX S INDIANA AVE  1506         PROSTITUTION   
15        019XX W MADISON ST  1121   DECEPTIVE PRACTICE   
16         047XX W FULTON ST  0486              BATTERY   
17          009XX W GRACE ST  0820                THEFT   
18         002XX N LOREL AVE  2027            NARCOTICS   
19          023XX S STATE ST  0460              BATTERY   
20       022XX N NATCHEZ AVE  1320      CRIMINAL DAMAGE   
21       133XX S BRANDON AVE  0460              BATTERY   
22     036XX S LAKE PA

In [5]:
# Starting the indexing from 1 so it can suffice for serial number
crime_chic.index += 1 

In [6]:
crime_chic.head(5)

Unnamed: 0        ID Case Number                    Date  \
1           0   6407111    HP485721  07/26/2008 02:30:00 PM   
2           1  11398199    JB372830  07/31/2018 10:57:00 AM   
3           2   5488785    HN308568  04/27/2007 10:30:00 AM   
4           3  11389116    JB361368  07/23/2018 08:55:00 AM   
5           4  12420431    JE297624  07/11/2021 06:40:00 AM   

                  Block  IUCR       Primary Type               Description  \
1  085XX S MUSKEGON AVE  1320    CRIMINAL DAMAGE                TO VEHICLE   
2     092XX S ELLIS AVE  143C  WEAPONS VIOLATION  UNLAWFUL POSS AMMUNITION   
3     062XX N TRIPP AVE  0610           BURGLARY            FORCIBLE ENTRY   
4    0000X N KEELER AVE  0560            ASSAULT                    SIMPLE   
5   016XX W HARRISON ST  051A            ASSAULT      AGGRAVATED - HANDGUN   

                     Location Description  Arrest  ...  Ward  Community Area  \
1                                  STREET   False  ...  10.0            46.0   
2                               POOL ROOM    True  ...   8.0            47.0   
3                               RESIDENCE    True  ...  39.0            12.0   
4            NURSING HOME/RETIREMENT HOME   False  ...  28.0            26.0   
5  PARKING LOT / GARAGE (NON RESIDENTIAL)   False  ...  27.0            28.0   

   FBI Code  X Coordinate  Y Coordinate  Year              Updated On  \
1        14     1196638.0     1848800.0  2008  02/28/2018 03:56:25 PM   
2        15     1184499.0     1843935.0  2018  08/07/2018 04:02:59 PM   
3        05     1146911.0     1941022.0  2007  02/28/2018 03:56:25 PM   
4       08A     1148388.0     1899882.0  2018  07/30/2018 03:52:24 PM   
5       04A     1165430.0     1897441.0  2021  07/18/2021 04:56:02 PM   

    Latitude  Longitude                       Location  
1  41.739980 -87.555120  (41.739979622, -87.555120042)  
2  41.726922 -87.599747  (41.726922145, -87.599746995)  
3  41.994138 -87.734959  (41.994137622, -87.734959049)  
4  41.881217 -87.730590  (41.881217483, -87.730589961)  
5  41.874174 -87.668082  (41.874173691, -87.668082118)  

[5 rows x 23 columns]

In [7]:
crime_chic.columns

Index(['Unnamed: 0', 'ID', 'Case Number', 'Date', 'Block', 'IUCR',
       'Primary Type', 'Description', 'Location Description', 'Arrest',
       'Domestic', 'Beat', 'District', 'Ward', 'Community Area', 'FBI Code',
       'X Coordinate', 'Y Coordinate', 'Year', 'Updated On', 'Latitude',
       'Longitude', 'Location'],
      dtype='object')

In [8]:
crime_chic.isna().sum()

Unnamed: 0                   0
ID                           0
Case Number                  1
Date                         0
Block                        0
IUCR                         0
Primary Type                 0
Description                  0
Location Description      2877
Arrest                       0
Domestic                     0
Beat                         0
District                    12
Ward                    184695
Community Area          184267
FBI Code                     0
X Coordinate             23985
Y Coordinate             23985
Year                         0
Updated On                   0
Latitude                 23985
Longitude                23985
Location                 23985
dtype: int64

In [9]:
#Dropping the 'Unnamed: 0' column
crime_chic.drop(['Unnamed: 0'], axis = 1, inplace = True)

In [10]:
#Checking for duplicates in the 'Case Number' column
crime_chic['Case Number'].duplicated().sum()

47

In [11]:
# Checking for duplicates in the 'Case Number' column
CaseNumber_duplicate = crime_chic[crime_chic.duplicated('Case Number')]

In [12]:
#Viewing the list of Case Number duplicates
CaseNumber_duplicate

ID Case Number                    Date                    Block  \
292312    3907    HM763064  12/08/2006 03:14:00 PM       005XX W MADISON ST   
322751   19393    HS313197  05/18/2010 06:37:00 AM          031XX W 48TH PL   
477866    2619    HK172551  02/08/2004 11:44:00 PM       034XX W MADISON ST   
502352    2620    HK172551  02/08/2004 11:44:00 PM       034XX W MADISON ST   
582265    4604    HP424011  06/29/2008 11:18:00 PM     078XX S KINGSTON AVE   
594161   22746    HZ403466  08/23/2016 01:34:00 AM        081XX S ESSEX AVE   
610917    2486    HJ756295  11/13/2003 09:25:00 AM        045XX W OGDEN AVE   
718479   20428    HV303630  05/25/2012 11:50:00 PM      065XX S KENWOOD AVE   
767268   24072    JB359724  07/21/2018 11:33:00 PM      031XX W FULTON BLVD   
1091023  22919    HZ497841  11/02/2016 04:45:00 PM          018XX W 58TH ST   
1165290  23517    JA389322  08/13/2017 11:45:00 AM      002XX S LARAMIE AVE   
1209811   3958    HN152307  01/30/2007 10:15:00 PM  010XX N SPRINGFIELD AVE   
1265601   3016    HL112637  01/08/2005 03:15:00 AM        015XX N DAYTON ST   
1286868   5008    HR300068  05/03/2009 04:33:00 PM          048XX N TROY ST   
1405827  20744    HV620483  12/29/2012 07:40:00 AM          042XX W 21ST PL   
1499445   1145     G614514  10/13/2001 01:44:00 AM        107XX S PERRY AVE   
1504828  22316    HZ140230  02/04/2016 01:00:00 PM   057XX S CALIFORNIA AVE   
1543494   2106    HJ309452  04/19/2003 01:48:00 AM          003XX E 51ST ST   
1564063  23190    JA185074  03/13/2017 04:44:00 PM        086XX S THROOP ST   
1588189  25263    JD286197  07/04/2020 11:57:00 PM     061XX S CARPENTER ST   
1608181   3489    HM154020  01/31/2006 09:00:00 PM       011XX N KEELER AVE   
1635304  21915    HY259141  06/10/2015 12:01:00 AM         002XX W 107TH ST   
1648800  26787    JF259064  05/28/2022 05:33:00 PM        055XX S BISHOP ST   
1668057   2119    HJ320635  04/24/2003 03:50:00 PM       062XX S PAULINA ST   
1743778  25311    JD298408  07/15/2020 04:11:00 PM        069XX S THROOP ST   
1768822   1961    HJ104730  01/03/2003 05:40:00 PM     080XX S MICHIGAN AVE   
1770308  26004    JE266473  06/15/2021 07:55:00 AM        062XX S MORGAN ST   
1804676  26628    JF152198  02/22/2022 09:12:00 AM   049XX S MARSHFIELD AVE   
1807673  24246    JB483810  10/21/2018 03:09:00 AM     001XX N LOCKWOOD AVE   
1858635  22541    HZ303764  06/12/2016 01:45:00 AM   090XX S MARSHFIELD AVE   
1967237   1958    HJ104730  01/03/2003 07:45:00 PM     080XX S MICHIGAN AVE   
1969091  19695    HS638640  11/30/2010 12:51:00 AM         111XX S BELL AVE   
1989014   4912    HR146859  01/31/2009 01:10:00 AM        039XX W ARGYLE ST   
2055747  23580    JA432970  09/15/2017 10:18:00 PM    047XX S FAIRFIELD AVE   
2084446  25920    JE230612  05/15/2021 06:34:00 AM      078XX S LOOMIS BLVD   
2088482  20139    HT576362  11/05/2011 05:05:00 PM          004XX E 87TH ST   
2092474  26243    JE344474  08/30/2021 09:17:00 PM       051XX W MADISON ST   
2093777   1234     G703238  11/23/2001 08:30:00 AM     112XX S GREEN STREET   
2147106  20473    HV341781  06/19/2012 01:12:00 AM        012XX N MASON AVE   
2150060  23208    JA206252  03/30/2017 04:32:00 PM          027XX E 75TH ST   
2185921  26701    JF203921  04/10/2022 11:20:00 PM     010XX W MONTROSE AVE   
2196936   2108    HJ309480  04/19/2003 06:10:00 AM        000XX S MASON AVE   
2211630  21120    HW486725  10/09/2013 09:43:00 PM      067XX S CHAPPEL AVE   
2218948   2350    HJ590004  08/27/2003 08:35:00 AM       039XX S WALLACE ST   
2223266   4501    HP296582  04/23/2008 06:15:00 PM       076XX S RHODES AVE   
2248272  21021    HW405390  08/13/2013 12:15:00 AM      028XX S KILDARE AVE   
2264226  20423    HV295764  05/21/2012 02:00:00 AM       022XX S LEAVITT ST   

         IUCR Primary Type          Description     Location Description  \
292312   0110     HOMICIDE  FIRST DEGREE MURDER                   OFFICE   
322751   0110     HOMICIDE  FIRST DEGREE MURDER                    

In [13]:
#Checking for duplicates in the 'ID' column
crime_chic['ID'].duplicated().sum()

0

OBSERVATION: There incidents with the same 'Case number'. These incident are all homicidal in nature, described as first degree murder, have the same 'IUCR' - 0110 and 'FBI code' - 01A

To this effect the column 'Case Number' will be dropped so each incident can be referenced with its 'ID'

In [14]:
#Dropping the 'Case number' column
crime_chic.drop(['Case Number'], axis = 1, inplace = True)

In [15]:
def missing_value_percentage(df):
    # Calculate the percentage of missing values for each column
    missing_percentage = df.isnull().mean().reset_index()
    missing_percentage.columns = ['Column', 'Missing Percentage']

#Converting the proportion to percentage
    missing_percentage['Missing Percentage'] *= 100

    return missing_percentage


In [16]:
missing_value_percentage(crime_chic)

Column  Missing Percentage
0                     ID            0.000000
1                   Date            0.000000
2                  Block            0.000000
3                   IUCR            0.000000
4           Primary Type            0.000000
5            Description            0.000000
6   Location Description            0.126255
7                 Arrest            0.000000
8               Domestic            0.000000
9                   Beat            0.000000
10              District            0.000527
11                  Ward            8.105187
12        Community Area            8.086404
13              FBI Code            0.000000
14          X Coordinate            1.052562
15          Y Coordinate            1.052562
16                  Year            0.000000
17            Updated On            0.000000
18              Latitude            1.052562
19             Longitude            1.052562
20              Location            1.052562

OBSERVATION: Going by the percentage of missing values, it is safe to say that there negligible as it is very minimal.

Further columns will be dropped as they aren't neccesarilly needed in our analysis. They are:
1. 'Ward' and 'Community Area' : these two columns plays a less significant role in our analysis since have the 'Location'. if such information is needed, it can easily be traced to the 'District' and from there the we can locate the 'Ward' and 'Community Area' via a map or the district data. In order words, the 'District' mothered the the Ward and 'Community Area'.
2. 'X and Y Coordinates' : these two column will give us location of the crime, same as what the 'Location' column will give us.
3. 'Latitute' and 'Longitude' : these two columns makes up the 'Location' column with the format - (Latitude, longitude).


In [17]:
#Dropping further columns
crime_chic.drop(['Ward','Community Area','X Coordinate', 'Y Coordinate','Latitude', 'Longitude'], axis = 1, inplace = True)

In [18]:
#Changing the column 'ID' header to 'Case ID'
crime_chic.rename(columns={'ID': 'Case ID'}, inplace=True, errors='raise')

In [19]:
crime_chic.head(5)

Case ID                    Date                 Block  IUCR  \
1   6407111  07/26/2008 02:30:00 PM  085XX S MUSKEGON AVE  1320   
2  11398199  07/31/2018 10:57:00 AM     092XX S ELLIS AVE  143C   
3   5488785  04/27/2007 10:30:00 AM     062XX N TRIPP AVE  0610   
4  11389116  07/23/2018 08:55:00 AM    0000X N KEELER AVE  0560   
5  12420431  07/11/2021 06:40:00 AM   016XX W HARRISON ST  051A   

        Primary Type               Description  \
1    CRIMINAL DAMAGE                TO VEHICLE   
2  WEAPONS VIOLATION  UNLAWFUL POSS AMMUNITION   
3           BURGLARY            FORCIBLE ENTRY   
4            ASSAULT                    SIMPLE   
5            ASSAULT      AGGRAVATED - HANDGUN   

                     Location Description  Arrest  Domestic  Beat  District  \
1                                  STREET   False     False   423       4.0   
2                               POOL ROOM    True     False   413       4.0   
3                               RESIDENCE    True     False  1711      17.0   
4            NURSING HOME/RETIREMENT HOME   False     False  1115      11.0   
5  PARKING LOT / GARAGE (NON RESIDENTIAL)   False     False  1231      12.0   

  FBI Code  Year              Updated On                       Location  
1       14  2008  02/28/2018 03:56:25 PM  (41.739979622, -87.555120042)  
2       15  2018  08/07/2018 04:02:59 PM  (41.726922145, -87.599746995)  
3       05  2007  02/28/2018 03:56:25 PM  (41.994137622, -87.734959049)  
4      08A  2018  07/30/2018 03:52:24 PM  (41.881217483, -87.730589961)  
5      04A  2021  07/18/2021 04:56:02 PM  (41.874173691, -87.668082118)

In [20]:
crime_chic.columns

Index(['Case ID', 'Date', 'Block', 'IUCR', 'Primary Type', 'Description',
       'Location Description', 'Arrest', 'Domestic', 'Beat', 'District',
       'FBI Code', 'Year', 'Updated On', 'Location'],
      dtype='object')

In [21]:
crime_chic.shape

(2278726, 15)

In [22]:
crime_chic.isna().sum()

Case ID                     0
Date                        0
Block                       0
IUCR                        0
Primary Type                0
Description                 0
Location Description     2877
Arrest                      0
Domestic                    0
Beat                        0
District                   12
FBI Code                    0
Year                        0
Updated On                  0
Location                23985
dtype: int64

In [23]:
crime_chic.District.unique()

array([ 4., 17., 11., 12.,  2., 15., 25., 14.,  8.,  6., 19.,  1.,  7.,
       20., 18.,  3.,  9., 22.,  5., 16., 10., 24., 31., nan])

In [24]:
# district 4 location values
crime_chic.groupby('District').get_group(4)[['Location']]

Location
1        (41.739979622, -87.555120042)
2        (41.726922145, -87.599746995)
22        (41.653220991, -87.54591594)
26       (41.752607015, -87.565034748)
43       (41.754533775, -87.565074073)
...                                ...
2278535  (41.747483166, -87.561314066)
2278561  (41.747961169, -87.576651771)
2278654  (41.725766223, -87.600518444)
2278717    (41.73327429, -87.55133578)
2278720  (41.732445186, -87.588037377)

[129838 rows x 1 columns]

In [25]:
# district 10 location values
crime_chic.groupby('District').get_group(10)[['Location']]

Location
84       (41.854498094, -87.723580391)
102       (41.852650452, -87.71624588)
118      (41.837823658, -87.729239549)
126      (41.847609583, -87.711222061)
149      (41.854564542, -87.712421035)
...                                ...
2278657   (41.862011728, -87.69378445)
2278660                            NaN
2278671  (41.858194929, -87.720032279)
2278673  (41.834442083, -87.721858991)
2278692  (41.841963269, -87.713508064)

[98172 rows x 1 columns]

In [33]:
#crime_chic.tail(2278659:2278668)
crime_chic.iloc[2278659:2278664]

Case ID                    Date                Block  IUCR  \
2278660  12773616  12/28/2021 12:00:00 AM      028XX W 22ND PL  2825   
2278661  11836584  09/21/2019 06:35:00 PM   062XX S PULASKI RD  0460   
2278662   2420378  10/24/2002 07:00:00 PM  032XX N LINCOLN AVE  0610   
2278663  10375394  01/11/2016 11:55:00 AM  062XX S ASHLAND AVE  0810   
2278664   3842076  03/03/2005 02:38:00 PM   038XX S WABASH AVE  0560   

          Primary Type              Description Location Description  Arrest  \
2278660  OTHER OFFENSE  HARASSMENT BY TELEPHONE            APARTMENT   False   
2278661        BATTERY                   SIMPLE              CTA BUS   False   
2278662       BURGLARY           FORCIBLE ENTRY       CLEANING STORE   False   
2278663          THEFT                OVER $500               STREET   False   
2278664        ASSAULT                   SIMPLE            RESIDENCE   False   

         Domestic  Beat  District FBI Code  Year              Updated On  \
2278660      True  1033      10.0       26  2021  07/27/2022 04:51:55 PM   
2278661     False   813       8.0      08B  2019  09/28/2019 04:01:29 PM   
2278662     False  1924      19.0       05  2002  02/28/2018 03:56:25 PM   
2278663     False   713       7.0       06  2016  02/10/2018 03:50:01 PM   
2278664      True   211       2.0      08A  2005  02/28/2018 03:56:25 PM   

                              Location  
2278660                            NaN  
2278661  (41.780242026, -87.722803147)  
2278662  (41.940852871, -87.669473528)  
2278663  (41.780835284, -87.664252074)  
2278664  (41.825048022, -87.624831462)

In [34]:
missing_rows = crime_chic[crime_chic['District'].isnull()]

In [35]:
missing_rows

Case ID                    Date                    Block  IUCR  \
249957   6514077  09/21/2008 10:00:00 PM       061XX S RHODES AVE  0550   
294135   6534600  10/03/2008 12:50:00 PM          039XX W 26TH ST  1130   
465266   6580042  10/21/2008 08:15:00 PM       0000X S KENTON AVE  1506   
621363   6580757  10/26/2008 03:02:00 AM          006XX W 66TH ST  0486   
869581   6579790  10/28/2008 12:30:00 PM       002XX E ONTARIO ST  0890   
889672   6594103  11/04/2008 08:14:29 PM   065XX N CALIFORNIA AVE  1320   
1024727  4740376  05/10/2006 07:40:00 PM      006XX E RANDOLPH ST  0460   
1215470  6588015  11/01/2008 04:10:00 PM        015XX N AVERS AVE  0840   
1868266  6405961  11/16/2007 12:01:00 AM           101XX S MAY ST  1150   
1929898  6488159  09/09/2008 05:00:00 PM     047XX N KEYSTONE AVE  0910   
2030068  6376239  07/18/2008 08:00:00 PM  010XX N SPRINGFIELD AVE  1310   
2237620  6516041  09/22/2008 09:10:00 PM   013XX N CALIFORNIA AVE  0560   

                Primary Type                    Description  \
249957               ASSAULT         AGGRAVATED PO: HANDGUN   
294135    DECEPTIVE PRACTICE       FRAUD OR CONFIDENCE GAME   
465266          PROSTITUTION          SOLICIT ON PUBLIC WAY   
621363               BATTERY        DOMESTIC BATTERY SIMPLE   
869581                 THEFT                  FROM BUILDING   
889672       CRIMINAL DAMAGE                     TO VEHICLE   
1024727              BATTERY                         SIMPLE   
1215470                THEFT  FINANCIAL ID THEFT: OVER $300   
1868266   DECEPTIVE PRACTICE              CREDIT CARD FRAUD   
1929898  MOTOR VEHICLE THEFT                     AUTOMOBILE   
2030068      CRIMINAL DAMAGE                    TO PROPERTY   
2237620              ASSAULT                         SIMPLE   

        Location Description  Arrest  Domestic  Beat  District FBI Code  Year  \
249957                 ALLEY    True     False   313       NaN      04A  2008   
294135                  BANK   False     False  1013       NaN       11  2008   
465266                STREET    True     False  1113       NaN       16  2008   
621363                STREET   False      True   723       NaN      08B  2008   
869581            RESTAURANT   False     False  1834       NaN       06  2008   
889672                 OTHER   False     False  2412       NaN       14  2008   
1024727                OTHER   False     False   124       NaN      08B  2006   
1215470            APARTMENT   False     False  2535       NaN       06  2008   
1868266            RESIDENCE   False      True  2232       NaN       11  2007   
1929898               STREET   False     False  1722       NaN       07  2008   
2030068            APARTMENT   False     False  1112       NaN       14  2008   
2237620             SIDEWALK   False     False  1423       NaN      08A  2008   

                     Updated On Location  
249957   02/28/2018 03:56:25 PM      NaN  
294135   02/28/2018 03:56:25 PM      NaN  
465266   02/28/2018 03:56:25 PM      NaN  
621363   02/28/2018 03:56:25 PM      NaN  
869581   02/28/2018 03:56:25 PM      NaN  
889672   02/28/2018 03:56:25 PM      NaN  
1024727  02/28/2018 03:56:25 PM      NaN  
1215470  02/28/2018 03:56:25 PM      NaN  
1868266  02/28/2018 03:56:25 PM      NaN  
1929898  02/28/2018 03:56:25 PM      NaN  
2030068  02/10/2018 03:50:01 PM      NaN  
2237620  02/10/2018 03:50:01 PM      NaN

In [40]:
#Filling the 'nan' value in the 'District' column with '0'
crime_chic['District'] = crime_chic['District'].fillna(0)

In [42]:
crime_chic.District.unique()

array([ 4., 17., 11., 12.,  2., 15., 25., 14.,  8.,  6., 19.,  1.,  7.,
       20., 18.,  3.,  9., 22.,  5., 16., 10., 24., 31.,  0.])

In [46]:
#Filling the 'nan' value in the 'Location Description' column with 'Unknown'
crime_chic['Location Description'] = crime_chic['Location Description'].fillna('Unknown')

In [ ]:
#first_missing_index = df['column_name'].index[df['column_name'].isnull()].tolist()[0]

In [44]:
crime_chic['Location'] = crime_chic.groupby('District')['Location'].transform(lambda x: x.fillna(x.mode()))

In [47]:
crime_chic.isna().sum()

Case ID                     0
Date                        0
Block                       0
IUCR                        0
Primary Type                0
Description                 0
Location Description        0
Arrest                      0
Domestic                    0
Beat                        0
District                    0
FBI Code                    0
Year                        0
Updated On                  0
Location                23997
dtype: int64

In [48]:
crime_chic.shape

(2278726, 15)

In [49]:
crime_chic.dropna(subset=['Location'], inplace=True)

In [50]:
crime_chic.shape

(2254729, 15)

In [51]:
crime_chic.isna().sum()

Case ID                 0
Date                    0
Block                   0
IUCR                    0
Primary Type            0
Description             0
Location Description    0
Arrest                  0
Domestic                0
Beat                    0
District                0
FBI Code                0
Year                    0
Updated On              0
Location                0
dtype: int64

In [53]:
# creating a column for months

from datetime import datetime

def extract_and_map_month(date):
    # ensure the input is in datetime format
    # 07/26/2008 02:30:00 PM
    date_column = pd.to_datetime(date, format='%m/%d/%Y %I:%M:%S %p')

    month_names = date_column.dt.month_name()

    return month_names

In [54]:
crime_chic['Month'] = extract_and_map_month(crime_chic['Date'])

In [55]:
crime_chic.head()

Case ID                    Date                 Block  IUCR  \
1   6407111  07/26/2008 02:30:00 PM  085XX S MUSKEGON AVE  1320   
2  11398199  07/31/2018 10:57:00 AM     092XX S ELLIS AVE  143C   
3   5488785  04/27/2007 10:30:00 AM     062XX N TRIPP AVE  0610   
4  11389116  07/23/2018 08:55:00 AM    0000X N KEELER AVE  0560   
5  12420431  07/11/2021 06:40:00 AM   016XX W HARRISON ST  051A   

        Primary Type               Description  \
1    CRIMINAL DAMAGE                TO VEHICLE   
2  WEAPONS VIOLATION  UNLAWFUL POSS AMMUNITION   
3           BURGLARY            FORCIBLE ENTRY   
4            ASSAULT                    SIMPLE   
5            ASSAULT      AGGRAVATED - HANDGUN   

                     Location Description  Arrest  Domestic  Beat  District  \
1                                  STREET   False     False   423       4.0   
2                               POOL ROOM    True     False   413       4.0   
3                               RESIDENCE    True     False  1711      17.0   
4            NURSING HOME/RETIREMENT HOME   False     False  1115      11.0   
5  PARKING LOT / GARAGE (NON RESIDENTIAL)   False     False  1231      12.0   

  FBI Code  Year              Updated On                       Location  Month  
1       14  2008  02/28/2018 03:56:25 PM  (41.739979622, -87.555120042)   July  
2       15  2018  08/07/2018 04:02:59 PM  (41.726922145, -87.599746995)   July  
3       05  2007  02/28/2018 03:56:25 PM  (41.994137622, -87.734959049)  April  
4      08A  2018  07/30/2018 03:52:24 PM  (41.881217483, -87.730589961)   July  
5      04A  2021  07/18/2021 04:56:02 PM  (41.874173691, -87.668082118)   July

In [59]:
# creating a column for seasons

def create_seasons_column(date_column):
    """
    Creates a new column representing seasons based on the months in a datetime column.

    Parameters:
    - date_column: Pandas Series, representing a datetime feature

    Returns:
    - Pandas Series containing the corresponding season names
    """
    # Ensure the input is in datetime format
    date_column = pd.to_datetime(date_column, format='%m/%d/%Y %I:%M:%S %p', errors='coerce')

    # Define a mapping of month to season
    month_to_season = {
        1: 'Winter', 2: 'Winter', 3: 'Spring',
        4: 'Spring', 5: 'Spring', 6: 'Summer',
        7: 'Summer', 8: 'Summer', 9: 'Autumn',
        10: 'Autumn', 11: 'Autumn', 12: 'Winter'
    }

    # Extract the month and map to season
    seasons_column = date_column.dt.month.map(month_to_season)

    return seasons_column

In [60]:
crime_chic['Seasons'] = create_seasons_column(crime_chic['Date'])

In [61]:
crime_chic.head()

Case ID                    Date                 Block  IUCR  \
1   6407111  07/26/2008 02:30:00 PM  085XX S MUSKEGON AVE  1320   
2  11398199  07/31/2018 10:57:00 AM     092XX S ELLIS AVE  143C   
3   5488785  04/27/2007 10:30:00 AM     062XX N TRIPP AVE  0610   
4  11389116  07/23/2018 08:55:00 AM    0000X N KEELER AVE  0560   
5  12420431  07/11/2021 06:40:00 AM   016XX W HARRISON ST  051A   

        Primary Type               Description  \
1    CRIMINAL DAMAGE                TO VEHICLE   
2  WEAPONS VIOLATION  UNLAWFUL POSS AMMUNITION   
3           BURGLARY            FORCIBLE ENTRY   
4            ASSAULT                    SIMPLE   
5            ASSAULT      AGGRAVATED - HANDGUN   

                     Location Description  Arrest  Domestic  Beat  District  \
1                                  STREET   False     False   423       4.0   
2                               POOL ROOM    True     False   413       4.0   
3                               RESIDENCE    True     False  1711      17.0   
4            NURSING HOME/RETIREMENT HOME   False     False  1115      11.0   
5  PARKING LOT / GARAGE (NON RESIDENTIAL)   False     False  1231      12.0   

  FBI Code  Year              Updated On                       Location  \
1       14  2008  02/28/2018 03:56:25 PM  (41.739979622, -87.555120042)   
2       15  2018  08/07/2018 04:02:59 PM  (41.726922145, -87.599746995)   
3       05  2007  02/28/2018 03:56:25 PM  (41.994137622, -87.734959049)   
4      08A  2018  07/30/2018 03:52:24 PM  (41.881217483, -87.730589961)   
5      04A  2021  07/18/2021 04:56:02 PM  (41.874173691, -87.668082118)   

   Month Seasons  
1   July  Summer  
2   July  Summer  
3  April  Spring  
4   July  Summer  
5   July  Summer

In [62]:
crime_chic.isna().sum()

Case ID                 0
Date                    0
Block                   0
IUCR                    0
Primary Type            0
Description             0
Location Description    0
Arrest                  0
Domestic                0
Beat                    0
District                0
FBI Code                0
Year                    0
Updated On              0
Location                0
Month                   0
Seasons                 0
dtype: int64